serdes-native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```diff
@@ -0,0 +1,360 @@
+     * Creates an anydata object from a byte array after deserializing.
+     *
+     * @param des  Deserializer object.
+     * @param encodedMessage Byte array corresponding to encoded data.
+     * @param dataType Data type of the encoded value.
+     * @return anydata object.
+     */
+    public static Object deserialize(BObject des, BArray encodedMessage, BTypedesc dataType) {
+        Descriptor schema = (Descriptor) des.getNativeData(SCHEMA_NAME);
+
+        Object object = null;
+        
+        try {
+            DynamicMessage dynamicMessage = generateDynamicMessageFromBytes(schema, encodedMessage);
+
+            object = dynamicMessageToBallerinaType(dynamicMessage, dataType, schema);
+        } catch (BError e) {
+            return e;
+        } catch (InvalidProtocolBufferException e) {
+            return createSerdesError(DESERIALIZATION_ERROR_MESSAGE + e.getMessage(), SERDES_ERROR);
+        }
+
+        return object;
+    }
+
+    private static DynamicMessage generateDynamicMessageFromBytes(Descriptor schema, BArray encodedMessage)
+                                                                                throws InvalidProtocolBufferException {
+        return DynamicMessage.parseFrom(schema, encodedMessage.getBytes());
+    }
+
+    private static Object dynamicMessageToBallerinaType(DynamicMessage dynamicMessage, BTypedesc typedesc,
+                                                        Descriptor schema) {
+        Type type = typedesc.getDescribingType();
+
+        if (type.getTag() <= TypeTags.BOOLEAN_TAG) {
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ATOMIC_FIELD_NAME);
+
+            return primitiveToBallerina(dynamicMessage.getField(fieldDescriptor));
+        } else if (type.getTag() == TypeTags.UNION_TAG) {
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ATOMIC_FIELD_NAME);
+            DynamicMessage dynamicMessageForUnion = (DynamicMessage) dynamicMessage.getField(fieldDescriptor);
+
+            return resolveUnionType(dynamicMessageForUnion, typedesc.getDescribingType(), schema);
+        } else if (type.getTag() == TypeTags.ARRAY_TAG) {
+            ArrayType arrayType = (ArrayType) type;
+            Type elementType = arrayType.getElementType();
+
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ARRAY_FIELD_NAME);
+            schema = fieldDescriptor.getContainingType();
+
+            return arrayToBallerina(dynamicMessage.getField(fieldDescriptor), elementType, schema, 1);
+        } else if (type.getTag() == TypeTags.RECORD_TYPE_TAG) {
+            Map<String, Object> mapObject = recordToBallerina(dynamicMessage, type, schema);
+
+            return ValueCreator.createRecordValue(type.getPackage(), type.getName(), mapObject);
+        } else {
+            throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+        }
+    }
+
+    private static Object primitiveToBallerina(Object value) {
```

Comment:
These method names are not clear enough. 

`getBallerinaPrimitiveValueFromNativeObject()` would be clearer. (This name is just an example, you can come up with an even better name). 

Check all the similar method names as well.